blockchain/process_test.go

```diff
@@ -138,13 +147,1025 @@ func TestProcessOrder(t *testing.T) {
+	// Ensure that a header that has a positional failure is rejected as
+	// expected.  Since invalid headers are not added to the block index,
+	// attempting to process it again is expected to return the specific failure
+	// reason as opposed to a known invalid block error.
+	//
+	//   ... -> bbm#
+	//              \-> b1badhdra
+	// -------------------------------------------------------------------------
+
+	g.RejectHeader("b1badhdra", ErrBadBlockHeight)
+	g.RejectHeader("b1badhdra", ErrBadBlockHeight)
+	g.ExpectBestInvalidHeader("bfbbadchild")
+
+	// -------------------------------------------------------------------------
+	// Ensure both headers and blocks whose parent header is not known are
+	// rejected as expected.
+	//
+	//   ... -> bbm# -> b1 (neither header nor block data known to chain)
+	//                    \-> b2
+	// -------------------------------------------------------------------------
+
+	g.RejectHeader("b2", ErrMissingParent)
+	g.RejectBlock("b2", ErrMissingParent)
+	g.ExpectBestInvalidHeader("bfbbadchild")
+
+	// -------------------------------------------------------------------------
+	// Ensure that the first few headers in the test chain that build on the
+	// final base maturity block are accepted in order.
+	//
+	//   ... -> bbm# -> b1 -> b2 -> b3
+	// -------------------------------------------------------------------------
+
+	g.AcceptHeader("b1")
+	g.AcceptHeader("b2")
+	g.AcceptHeader("b3")
+	g.ExpectBestHeader("b3")
+
+	// -------------------------------------------------------------------------
+	// Ensure that headers that form a block index with a bunch of branches from
+	// different fork points are accepted and result in the one with the most
+	// cumulative work being identified as the best header that is not already
+	// known to be invalid.
+	//
+	// Note that the ! below indicates a block that is invalid due to violating
+	// a consensus rule, but has a valid header.
+	//
+	//  ... -> b3 -> b4  -> b5   -> b6  -> b7  -> b8  -> b9  -> b10  -> b11
+	//           |                                         \-> b10a
+	//           \-> b4b -> b5b! -> b6b -> b7b -> b8b
+	//           \-> b4c               |      \-> b8c -> b9c
+	//                                 \-> b7d -> b8d -> b9d
+	//                                 \-> b7e
+	// -------------------------------------------------------------------------
+
+	g.AcceptHeader("b4")
+	g.AcceptHeader("b4b")
+	g.AcceptHeader("b4c")
+	g.AcceptHeader("b5")
+	g.AcceptHeader("b5b") // Invalid block, but header valid.
+	g.AcceptHeader("b6")
+	g.AcceptHeader("b6b")
+	g.AcceptHeader("b7")
+	g.AcceptHeader("b7b")
+	g.AcceptHeader("b7d")
+	g.AcceptHeader("b7e")
+	g.AcceptHeader("b8")
+	g.AcceptHeader("b8b")
+	g.AcceptHeader("b8c")
+	g.AcceptHeader("b8d")
+	g.AcceptHeader("b9")
+	g.AcceptHeader("b9c")
+	g.AcceptHeader("b9d")
+	g.AcceptHeader("b10")
+	g.AcceptHeader("b10a")
+	g.AcceptHeader("b11")
+	g.ExpectBestHeader("b11")
```

Comment:
I noticed you're not testing best header selection the same you way you test best invalid header selection and best tip selection.

Do you think it's worth it to add such assertions (for example, by accepting the b#b branch first, then ensuring that accepting the b# branch later correctly selects b11 as best header) or are the tests added for `workSorterLess()` sufficient?

Reply:
I added a bunch of best header tests (both best and best invalid) as well as is current tests in the invalidate/reconsider PR, so I didn't add too many here, but I can go back and add some here too.  It never hurts to further validate assumptions.